src/libmscore/draw/drawcomp.cpp

```diff
@@ -0,0 +1,443 @@
+//=============================================================================
+//  MuseScore
+//  Music Composition & Notation
+//
+//  Copyright (C) 2021 MuseScore BVBA and others
+//
+//  This program is free software; you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License version 2.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program; if not, write to the Free Software
+//  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+//=============================================================================
+#include "drawcomp.h"
+
+#include "realfn.h"
+#include "log.h"
+
+using namespace mu;
+using namespace mu::draw;
+
+namespace mu::draw::comp {
+static const int DEFAULT_PREC(3);
+
+template<class T>
+static bool isEqual(const std::vector<T>& v1, const std::vector<T>& v2, DrawComp::Tolerance tolerance);
+
+static bool isEqual(const qreal& v1, const qreal& v2, double tolerance)
+{
+    if (tolerance > 0) {
+        double delta = RealFloor(v1, DEFAULT_PREC) - RealFloor(v2, DEFAULT_PREC);
+        return std::fabs(delta) < tolerance;
+    }
+    return RealIsEqual(RealFloor(v1, DEFAULT_PREC), RealFloor(v2, DEFAULT_PREC));
+}
+
+static bool isEqual(const QPointF& p1, const QPointF& p2, double tolerance)
+{
+    if (!isEqual(p1.x(), p2.x(), tolerance)) {
+        return false;
+    }
+
+    if (!isEqual(p1.y(), p2.y(), tolerance)) {
+        return false;
+    }
+
+    return true;
+}
+
+static bool isEqual(const QRectF& r1, const QRectF& r2, double tolerance)
+{
+    if (!isEqual(r1.x(), r2.x(), tolerance)) {
+        return false;
+    }
+
+    if (!isEqual(r1.y(), r2.y(), tolerance)) {
+        return false;
+    }
+
+    if (!isEqual(r1.width(), r2.width(), tolerance)) {
+        return false;
+    }
+
+    if (!isEqual(r1.height(), r2.height(), tolerance)) {
+        return false;
+    }
+
+    return true;
+}
+
+static bool isEqual(const QSize& sz1, const QSize& sz2)
+{
+    if (sz1.width() != sz2.width()) {
+        return false;
+    }
+
+    if (sz1.height() != sz2.height()) {
+        return false;
+    }
+
+    return true;
+}
+
+static bool isEqual(const QPen& p1, const QPen& p2)
```

Comment:
Can we just use == instead of the current implementation?

 static bool isEqual(const QPen& p1, const QPen& p2) {
 return p1 == p2;
 }

Reply:
There are a number of properties that I do not compare, for example brush, because I do not save them in json, respectively, I do not read them json, because they are not important to us, but because of them the comparison maybe failing.